ActiveRecord: do not create "has many through" records that have been removed

kratob opened the pull request:
### Summary

If a record was built on a HasManyThroughAssociation, then removed, and
then the record was saved, the removed record would be created anyways.

Code would look like this:


```
class Post
  #...
  has_many_people through: :readers
end

post = Post.new
post.people.build(first_name: "Bob")
ted = post.people.build(first_name: "Ted")
post.people.delete(ted)
# post.people now expectedly contains only 'Bob'

post.save!
post.reload
# post.people now unexpectedly contains 'Bob' and 'Ted'
```

This happens because records and not removed from the hidden "through association" on the join table, in the case of unpersisted child records.

The fix is to remove this records from the "through association".

kratob committed:
ActiveRecord: do not create "has many through" records that have been removed

If a record was built on a HasManyThroughAssociation, then removed, and
then the record was saved, the removed record would be created anyways.

rafaelfranca commented on activerecord/test/cases/associations/has_many_through_associations_test.rb:
```diff
@@ -319,6 +319,17 @@ def test_build_then_save_with_has_one_inverse
+    post.save!
+    post.reload
+
+    assert_equal ['Bob'], post.people.collect(&:first_name)
```
Could you use double quotes here?

rafaelfranca reviewed (changes requested):
Can you add a CHANGLOG entry?

kratob commented:
Done.

eileencodes commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+*   When using `has_many_through`, building records on the association, then deleting them
+    and then saving the parent record will no longer save the previously deleted records.
+
+    *Tobias Kraze*
```
`has_many_through` should be `has_many :through`. I also think the sentence is a little bit of a run on and would read clearer as:

> Previously when building records using a `has_many :through` association, if the child records were deleted before the parent was saved they would be persisted. Now if child records are deleted before the parent is saved on a `has_many :through` association the child records will now be deleted.

eileencodes reviewed (changes requested):
Looks good, I requested a change to the change log.

kratob commented:
I've updated the change log with a slight modification to your proposal. I changed "will be deleted" to "will not be persisted", as those records had never been persisted in the first place.

eileencodes commented:
Thanks for working on this @kratob!
